Adding traceheaders

achaikou opened the pull request:
Supports opening files with more than just standard traceheader.
Data is available internally in python, but not for end users yet.
    
Full disclosure regarding matlab code changes: taken approach is "if CI job passes, we are good" :grin:

achaikou committed:
Deprecate ext_headers

Deprecated to avoid confusion.

Despite stanzas being new way of working with text headers (not yet
available for python user), we don't remove the option to interact
directly with raw headers.

achaikou committed:
Update segy_trace0 with revision 2.1 requirements

trace0 type is changed to definitely larger, value is attempted to be
read from corresponding field and if that is not available, correct
number of extended text headers is required.

achaikou committed:
Allow to read and write all traceheaders

Functionality is still mainly tested via standard header.

Endianness-swap is not tested besides already existing tests for
standard header. We do not have enough tools yet to create swapped files
with several trace headers.

We might decide to change this interface again if we would store all
mappings (not just standard and extension 1) in C code.

achaikou committed:
Expose traceheaders read and write to python

ajaust commented on lib/src/segy.c:
```diff
@@ -942,13 +942,20 @@ int segy_flush( segy_datasource* ds ) {
+    long long pos = trace0 +
+                    trace * trace_size +
+                    traceheader * SEGY_TRACE_HEADER_SIZE +
+                    offset;
```
Do we need to check if `traceheader` is smaller than the number of trace headers? I could see that we start seeking inside the trace samples without noticing. 

Example:
The file has 3 trace headers per trace (standard trace header + 2 extended trace headers) and 1000 samples. Calling with `seek_trace_header = 3` would seek within for the header within the samples of the current trace.

1. Could this ever happen?
2. Do we safeguard against this or is this considered a user error.

yngve793 commented on mex/segy_get_ps_line_mex.c:
```diff
@@ -26,19 +26,23 @@ void mexFunction(int nlhs, mxArray *plhs[],
     int il = (int)mxGetScalar(mx_il_word);
     int xl = (int)mxGetScalar(mx_xl_word);
 
-    segy_file* fp = segyio_open( spec.filename, "rb" );
```
Is this a bug fix or have we changed the name in this PR?

yngve793 replied:
OK, fine.

ajaust reviewed (commented):
The changes look great. 🙂 

I left some remark w.r.t. changed data type of `trace0` which might be critical. Most other comments are mainly for my understanding and clarify our assumptions.

yngve793 reviewed (approved):
It is nice to see that bits and pieces are starting to come together. 

This PR contains simpler and in general more basic changes. 
I have only added some minor comments.

achaikou committed:
Rename segy_seek to seek_traceheader_offset

To make code more readable segy_seek is renamed to
seek_traceheader_offset and adjusted accordingly.

achaikou committed:
Remove trace0 and trace_bsize from signatures

Now values are expected to be set on segy_metadata field.

segy_sample_interval logic has changed as it no longer retrieves trace0
position, but expects it to be set already.

achaikou committed:
Learn to read files with trace header extensions

trace-header-extensions.sgy file is created manually, and will be
updated manually when more features are needed. In the future it should
be possible to recreate it with a script.
In this commit it is enough that segyio can read it.

Binary applications are not updated and still work on revision 1 only.

achaikou reviewed (commented):
Thanks for the reviews! :smiley: 
Fixups are added.

I've already rebased changes to main and there were no controversial merge conflicts.

achaikou committed:
Expose traceheader_count to python

achaikou committed:
Rename segy_traceheader and segy_write_traceheader

yngve793 commented on lib/test/segy.cpp:
```diff
@@ -1400,20 +1358,18 @@ SCENARIO( "modifying trace header", "[c.segy]" ) {
 namespace {
 
 int check_sorting( const std::string& name ) {
-    int trace0 = 3600;
```
ok

yngve793 commented on python/segyio/segyio.cpp:
```diff
@@ -1059,9 +1053,7 @@ PyObject* getth( segyfd* self, PyObject *args ) {
                            SEGY_TRACE_HEADER_SIZE, buffer.len() );
 
     int err = segy_traceheader( ds, traceno,
```
No, need then.

ajaust commented on lib/include/segyio/segy.h:
```diff
@@ -563,8 +563,20 @@ int segy_trace_bsize( int samples );
+ * calculate implied trace0 consulting ext_textheader_count. If provided value
+ * is negative, attempt would be made to read value from fields 3505–3506 and
+ * error code is returned if this value is negative too.
+ */
```
Thanks for the explanation. 🙂

ajaust reviewed (approved):
Looks good from my side. 👍
